spec/counters/workflow_counter_spec.rb

```diff
@@ -41,6 +41,20 @@
           expect(counter.classifications).to eq(2)
         end
       end
+
+      context "with subject sets marked as training data" do
+        let(:workflow) { create(:workflow_with_subjects, num_sets: 2) }
+        let(:training_set) { workflow.subject_sets.first }
+        let(:real_set) { workflow.subject_sets.last }
+        before do
+          real_set_ar_collection_proxy = SubjectSet.where(id: real_set)
+          allow(workflow).to receive(:non_training_subject_sets).and_return(real_set_ar_collection_proxy)
+        end
+
+        it "should return non training data classification count only" do
+          expect(counter.classifications).to eq(2)
```

Comment:
Is this counter using the workflow from the top of the spec or the new one with the same name you've built in this context? I assume it's not evaluated until you call #classifications so it uses the new workflow, but could you just add the training set to the existing workflow?

Reply:
> Is this counter using the workflow from the top of the spec or the new one with the same name you've built in this context?

In Rspec any nested let will override the outer let clauses, so this context `let(:workflow)` overrides the root `let(:workflow)`. When we build the counter in the root it will use the nested context let. 

> I assume it's not evaluated until you call #classifications so it uses the new workflow

As it's lazy loaded, i.e. the workflow will only get built when the counter is created via it's own let clause.

> but could you just add the training set to the existing workflow?

Yep but i feel that the workflow factory is easier to read vs manually adding a set to ensuring the correct relations and avoid building extra instances (like workflow and project)
```
let(:workflow) { create(:workflow_with_subjects, num_sets: 2) }
let(:training_set) { workflow.subject_sets.first }
let(:real_set) { workflow.subject_sets.last }
## vs
let(:training_set) do 
  create(:subject_set, project: workflow.project, workflows: [workflows]) do |set|
    workflow.subject_sets << set
  end
end
let(:real_set) { workflow.non_training_subject_sets.first }
before do
  # ensure the training set is setup on the workflow before the specs run
  training_set
  ....
end
```